src/types/osstr.rs

```diff
@@ -0,0 +1,191 @@
+use crate::types::PyString;
+#[cfg(not(windows))]
+use crate::PyNativeType;
+use crate::{
+    ffi, AsPyPointer, FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python,
+    ToPyObject,
+};
+use std::borrow::Cow;
+use std::ffi::{OsStr, OsString};
+#[cfg(not(windows))]
+use std::os::raw::c_char;
+
+impl ToPyObject for OsStr {
+    fn to_object(&self, py: Python) -> PyObject {
+        // If the string is UTF-8, take the quick and easy shortcut
+        if let Some(valid_utf8_path) = self.to_str() {
+            return PyString::new(py, valid_utf8_path).into();
+        }
+
+        // Take advantage of Python's surrogateescape encoding to support arbitrary byte
+        // sequences instead of limiting ourselves to UTF-8
+
+        // All targets besides windows support the std::os::unix::ffi::OsStrExt API:
+        // https://doc.rust-lang.org/src/std/sys_common/mod.rs.html#59
+        #[cfg(not(windows))]
+        {
+            let bytes = std::os::unix::ffi::OsStrExt::as_bytes(self);
+
+            let ptr = bytes.as_ptr() as *const c_char;
+            let len = bytes.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is this the appropriate function? Does it use the surrogateescape error
+                // handler? Should it be PyUnicode_DecodeLocaleAndSize instead maybe?
+                let pystring = ffi::PyUnicode_DecodeFSDefaultAndSize(ptr, len);
+                py.from_owned_ptr::<PyString>(pystring).into()
+            }
+        }
+
+        #[cfg(windows)]
+        {
+            let wstr: Vec<u16> = std::os::windows::ffi::OsStrExt::encode_wide(self).collect();
+
+            let ptr = wstr.as_ptr();
+            let len = wstr.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is there some better thing to do here than panic? Some last resort thing to
+                // try?
+                py.from_owned_ptr_or_err::<PyString>(ffi::PyUnicode_FromWideChar(ptr, len))
+                    .expect("OsStr conversion to Python string via FromWideChar failed")
+                    .into()
```

Comment:
I'd just use `PyObject::from_owned_ptr` here.

```suggestion
                PyObject::from_owned_ptr(py, ffi::PyUnicode_FromWideChar(ptr, len))
```

Reply:
I put the `.expect()` there as a sort of temporary measure, because I didn't know under what circumstances PyUnicode_FromWideChar would fail. There may be circumstances which we can handle appropriately.

Rust's encode_wide "re-encodes an OsStr as a wide character sequence, i.e., potentially ill-formed UTF-16." and PyUnicode_FromWideChar "creates a Unicode object from the wchar_t buffer w of the given size." The "potentially ill-formed" part trips me up. We don't want PyO3 to panic on regular usage, right?

Reply:
On second thought, the "potentially ill-formed" probably refers to the fact that Windows uses some unholy modified variant of UTF-16 internally. I'm pretty sure that the data emitted from encode_wide exactly compatible with PyUnicode_FromWideChar. So, no errors should happen and the hidden panic is fine (because it signifies some unforeseen error like a programming bug)